### ISSUE TITLE
Implement something like `torch.einsum` for `Tensor`s

### ISSUE THREAD

**Issue by TheLostLambda** [tracel-ai/burn#2480]
### Feature description

Something like https://pytorch.org/docs/stable/generated/torch.einsum.html, with something similar done in Rust already here: https://docs.rs/einsum-codegen/latest/einsum_codegen/

### Feature motivation

It's a more compact way to write out a number of tensor operations that come up throughout ML work.

### (Optional) Suggest a Solution

It seems like a good place for writing a macro! It looks like there are some details on the algorithm here: https://docs.rs/einsum-codegen/latest/einsum_codegen/#einsum-algorithm

**Comment by antimora** [comment]
@crutcher, Not sure if this falls under `linalg`, but just in case fyi

**Comment by BhavyeMathur** [comment]
Hey! I recently implemented einsums for my personal Rust ML library. The relevant code can be found [here](https://github.com/BhavyeMathur/redstone-ml/blob/master/src/linalg/einsum.rs#L403). Could I be assigned this issue?

My implementation was modelled after NumPy and works as follows:

1. Parse the `einsum` input and output arguments, keeping a track of unique indices, the count of each index, etc.
2. Check if a view into the tensor can be returned, and do so if possible
3. Check if any dimensions can be collapsed for faster iteration
4. If we have 2 or 3 iteration axes (+some required conditions), dispatch to a specialised einsum loop
5. Conduct an unspecialised einsum loop

In both the specialised and unspecialised loop, we make use of a 'sum of products' function selected according to the memory layout of the tensor. More can be found [here](https://github.com/BhavyeMathur/redstone-ml/blob/master/src/linalg/sum_of_products.rs). These loops are where we can provide GPU/CPU acceleration.

The architecture of Burn seems quite different and I'm unsure if I can just implement something as described above or if there are other things I must keep in mind as well (such as kernel fusion, the autodiff engine, etc). I'm quite new to Burn so any advice on how to go about implementing this would be greatly appreciated!

**Comment by laggui** [comment]
> Hey! I recently implemented einsums for my personal Rust ML library. The relevant code can be found [here](https://github.com/BhavyeMathur/redstone-ml/blob/master/src/linalg/einsum.rs#L403). Could I be assigned this issue?

Of course! I just assigned it to you. Feel free to remove the assignment if ever you change your mind 🙂 

> The architecture of Burn seems quite different and I'm unsure if I can just implement something as described above or if there are other things I must keep in mind as well (such as kernel fusion, the autodiff engine, etc). I'm quite new to Burn so any advice on how to go about implementing this would be greatly appreciated!

I would implement this as a high-level tensor operation. Maybe just [for float](<https://github.com/tracel-ai/burn/blob/main/crates/burn-tensor/src/tensor/api/float.rs>)? 

We can parse the einsum input/output args to lower it to a specific backend operation like matmul (`B::float_matmul`), reduction, etc. More complex patterns might be lowered to a sequence of these ops. Fusion supports elemwise ops, matmul and reductions so technically even a sequence of these operations with fusion will likely generate a single fused kernel.

**Comment by antimora** [comment]
einsum landed in #5583 (`burn-einsum`, with runtime and macro APIs). Closing.
